### PR TITLE
Simulated Drilling Machine and Conveyor Belt

### PR DESCRIPTION
**Pull request by moriarty**
@svenschneider - can you take a look, thanks.

### CODE REVIEW

File: cfg/config.yaml [hunk 1]
```diff
@@ -12,7 +12,8 @@ llsfrb:
     game: game.log
 
   conveyor-belt:
-    enable: false
+    #mode: simulation, real, mockup
```
Comment (svenschneider) [inline cfg/config.yaml:15]: For the other options we don't give the possible values. Therefore, I would not include them here. However, the CFH manual is probably a good place for them.

File: cfg/config.yaml [hunk 2]
```diff
@@ -57,9 +58,9 @@ llsfrb:
 
     public-peer:
       host: !ipv4 192.168.1.255
-      port: !udp-port 4444
```
Comment (svenschneider) [inline cfg/config.yaml:60]: I prefer having the "real" setup here (i.e. the old version).

File: cfg/config.yaml [hunk 3]
```diff
@@ -116,5 +117,5 @@ llsfrb:
       wits: randomkey
 
   shell:
-    refbox-host: !ipv4 192.168.1.100
```
Comment (svenschneider) [inline cfg/config.yaml:119]: Same as above.

File: rockin/plugins/conveyor_belt/conveyor_belt_plugin.cpp [hunk 4]
```diff
@@ -36,7 +36,25 @@ class ConveyorBeltPlugin: public fawkes::Plugin
         ConveyorBeltPlugin(Configuration *config) :
                 Plugin(config)
         {
-            thread_list.push_back(new ConveyorBeltThread());
+            std::string mode = "";
+            if (config->exists("/llsfrb/conveyor-belt/mode")) {
+                std::string config_mode = config->get_string("/llsfrb/conveyor-belt/mode");
+                if (config_mode == std::string("real"))
+                {
```
Comment (svenschneider) [inline rockin/plugins/conveyor_belt/conveyor_belt_plugin.cpp:43]: Put the brace to the end of the previous line, here.

File: rockin/plugins/conveyor_belt/conveyor_belt_plugin.cpp [hunk 5]
```diff
@@ -36,7 +36,25 @@ class ConveyorBeltPlugin: public fawkes::Plugin
...
+                    mode = "real";
```
Comment (svenschneider) [inline rockin/plugins/conveyor_belt/conveyor_belt_plugin.cpp:44]: I thought our discussion was to have different implementations of the ConveyorBeltThread? Then you would just instantiate the "correct" one here.

File: rockin/plugins/conveyor_belt/conveyor_belt_thread.cpp [hunk 6]
```diff
@@ -30,19 +30,21 @@
  */
 
 /** Constructor. */
-ConveyorBeltThread::ConveyorBeltThread() :
+ConveyorBeltThread::ConveyorBeltThread(std::string mode) :
         Thread("ConveyorBeltThread", Thread::OPMODE_CONTINUOUS), zmq_context_(NULL), zmq_publisher_(NULL), zmq_subscriber_(NULL), cfg_timer_interval_(40), default_network_interface_("eth0")
 {
+    mode_ = mode;
+    last_status_mutex_ = new fawkes::Mutex();
```
Comment (svenschneider) [inline rockin/plugins/conveyor_belt/conveyor_belt_thread.cpp:37]: Why don't you allocate the mutex on the stack (as a member in the class, but not a pointer)?

File: rockin/plugins/conveyor_belt/conveyor_belt_thread.cpp [hunk 7]
```diff
@@ -104,24 +123,42 @@ void ConveyorBeltThread::init()
...
 
 void ConveyorBeltThread::loop()
 {
-    receiveAndBufferStatusMsg();
+    if (mode_ == "real")
+    {
+        receiveAndBufferStatusMsg();
+    } else if (mode_ == "simulation")
+    {
+        fawkes::MutexLocker lock(last_status_mutex_);
```
Comment (svenschneider) [inline rockin/plugins/conveyor_belt/conveyor_belt_thread.cpp:145]: This looks "asymmetric", in the "real" case you don't have a lock, here you have one (_I_ know that the lock is inside the receiveAndBufferStatusMsg function, but someone else might wonder).

File: rockin/plugins/conveyor_belt/conveyor_belt_thread.cpp [hunk 8]
```diff
@@ -104,24 +123,42 @@ void ConveyorBeltThread::init()
...
+
+        if (!last_status_msg_.has_mode())
+        {
+            last_status_msg_.set_mode(STOP);
+            last_status_msg_.set_is_device_connected(true);
+            logger->log_info("ConveyorBelt", "simulation initialized to stop");
```
Comment (svenschneider) [inline rockin/plugins/conveyor_belt/conveyor_belt_thread.cpp:151]: Maybe debug is sufficient.

Comment (moriarty) [inline rockin/plugins/conveyor_belt/conveyor_belt_thread.cpp:151]: Oh that's a mistake- that was just for testing.

File: rockin/plugins/drilling_machine/drilling_machine_plugin.cpp [hunk 9]
```diff
@@ -36,8 +36,28 @@ class DrillingMachinePlugin: public fawkes::Plugin
         DrillingMachinePlugin(Configuration *config) :
                 Plugin(config)
         {
-            thread_list.push_back(new DrillingMachineThread());
-        }
+          std::string mode = "";
+          int encoder_max = 40;
```
Comment (svenschneider) [inline rockin/plugins/drilling_machine/drilling_machine_plugin.cpp:40]: Why is this here?

**Comment by svenschneider** [comment]
The comments for the conveyor belt apply also to the drilling machine.

File: rockin/plugins/conveyor_belt/conveyor_belt_thread.cpp (hunk 7 above)
Comment (moriarty) [inline rockin/plugins/conveyor_belt/conveyor_belt_thread.cpp:145]: Originally, I had all of the simulation inside of receiveAndBufferStatusMsg. 

But I moved it here because instead because it doesn't have anything to do with the functions name recieve and buffer status msgs.

File: rockin/plugins/drilling_machine/drilling_machine_plugin.cpp (hunk 9 above)
Comment (moriarty) [inline rockin/plugins/drilling_machine/drilling_machine_plugin.cpp:40]: This should be a configuration param or a constant. 

The problem I have is that the loop rate is a configuration param, so I can't know which rate the ticks_ are increasing. 

Which would you prefer?

**Comment by svenschneider** [comment]
Is there any update on this?

**Comment by moriarty** [comment]
Pull request #13 closes this.
